[Distributed] Protocol mangling and primary associated type fixes

ktoso opened the pull request:
The commits are now properly staged, please check their descriptions for details.

This solves:
- rdar://139332556 - so now we're handling primary associated types in protocols with @resolvable -- and this would have then crashed during mangling (!), which is why the follow up commits
- rdar://139781083 - which is solved in [`2a3ac44` (#77584)](https://github.com/swiftlang/swift/pull/77584/commits/2a3ac4458fcca51906ffd53a40129347ed8d81d7) which was about how sometimes with generics involved or more complex type signatures of a `distributed func` we'd wrongly mangle and crash during thunk mangling.

Both changes go hand in hand and allow correct handling of more situations with generics and complex types in `distributed func`s

xedin commented:
@ktoso I see that there is a bunch of leftover comments and code, could you please clean that up and ping me when done so I can take a look when this is ready to merge?

ktoso commented:
Oh yeah don't review this at all yet, I'm cleaning up and will make it into some nice commits.

ktoso commented:
@swift-ci please smoke test

ktoso committed:
[Distributed] @Resolvable now handles primary associated types in protocols

Previously we would not propagate those into the generated distributed
actor, making a lot of generic distributed actor protocols impossible to
express.

We indeed cannot handle protocols WITHOUT primary associated types, but
we certainly can handle them with!

This resolves rdar://139332556

ktoso committed:
[Distributed] Introduce SWIFT_DUMP_ACCESSIBLE_FUNCTIONS

ktoso committed:
[Distributed] More tests for primary associated types in protocols

ktoso committed:
[Distributed] Correct distributed accessor mangling for protocol calls

This is a crucial fix without which we can crash on some distributed
protocol declarations with @Resolvable. We cannot "just" use a String to
represent the "fake base" of the thunks, and must instead find the
$Target macro generated type and use it as the base of the thunk's
mangling.

Calls are made in such way that record for the protocol requirement:
`$s4main28GreeterDefinedSystemProtocolP5greetSSyYaKFTEHF` points at
`$$s4main29$GreeterDefinedSystemProtocolC5greetSSyYaKFTE` which makes a
dispatch through the _apropriate_ witness table.
And the record for the $witness named e.g.
`$s4main29$GreeterDefinedSystemProtocolC5greetSSyYaKFTEHF` points to
`$s4main28GreeterDefinedSystemProtocolPAA11Distributed01_F9ActorStubRzrlE5greetSSyYaKFTE`
which is an extension method: `distributed thunk (extension in main):main.GreeterDefinedSystemProtocol< where A: Distributed._DistributedActorStub>.greet() async throws -> Swift.String`,
this very specific design allows us to call the "right method" on the
recieving end of a remote call where we do not know the recipient type.

ktoso committed:
[Distributed] minor cleanups, remove dump=always from tests

ktoso commented:
@swift-ci please smoke test

ktoso committed:
[DistributedResolvableMacro] cannot use if expressions

ktoso commented:
@swift-ci please smoke test

ktoso commented:
Unrelated, Cxx interop fails compiling on Windows:

```
inalloca argument for call has mismatched alloca

  %7 = alloca %TSo3stdO0077basic_stringCChar32char_traitsCChar32allocatorCChar32_jlHEvsaGGdtaEBguasEitaaV, align 4

  %9 = call i32 @"?__swift_interopComputeHashOfU32String@@YAIV?$basic_string@_UU?$char_traits@_U@std@@V?$allocator@_U@2@@std@@@Z"(ptr %7) #16

<unknown>:0: error: fatal error encountered during compilation; please submit a bug report (https://swift.org/contributing/#reporting-bugs)

```

ktoso commented:
@swift-ci please smoke test Windows

ktoso commented on test/Distributed/Runtime/distributed_actor_remoteCall_protocol_method_in_presence_of_multiple_systems.swift:
```diff
@@ -46,16 +67,18 @@ distributed actor DAFL: GreeterProtocol {
+    // CHECK-NEXT: > encode return type: Swift.String
+    // CHECK-NEXT: > done recording
+    // CHECK-NEXT: >> remoteCall: on:main.$GreeterProtocol<main.FakeRoundtripActorSystem>, target:main.$GreeterProtocol.greet(), invocation:FakeInvocationEncoder(genericSubs: [main.$GreeterProtocol<main.FakeRoundtripActorSystem>], arguments: [], returnType: Optional(Swift.String), errorType: nil), throwing:Swift.Never, returning:Swift.String
+    // CHECK-NEXT: > execute distributed target: main.$GreeterProtocol.greet(), identifier: $s4main16$GreeterProtocolC5greetSSyYaKFTE
```
note: here we keep the same mangling, good 👍

ktoso commented on test/Distributed/Runtime/distributed_actor_func_calls_remoteCall_extension.swift:
```diff
@@ -40,7 +40,9 @@ func test() async throws {
   let reply = try await ref.echo(name: "Caplin")
-  // CHECK: >> remoteCall: on:main.KappaProtocolImpl, target:main.$KappaProtocol.echo(name:), invocation:FakeInvocationEncoder(genericSubs: [main.KappaProtocolImpl], arguments: ["Caplin"], returnType: Optional(Swift.String), errorType: nil), throwing:Swift.Never, returning:Swift.String
+  // CHECK: >> remoteCall: on:main.KappaProtocolImpl, target:echo(name:), invocation:FakeInvocationEncoder(genericSubs: [main.KappaProtocolImpl], arguments: ["Caplin"], returnType: Optional(Swift.String), errorType: nil), throwing:Swift.Never, returning:Swift.String
+  // CHECK: > execute distributed target: echo(name:), identifier: $s4main13KappaProtocolPAAE4echo4nameS2S_tYaKFTE
```
added a mangling check as well, we didnt before

ktoso commented:
@swift-ci please smoke test

xedin commented on lib/AST/ASTMangler.cpp:
```diff
@@ -4452,12 +4452,32 @@ void ASTMangler::appendDistributedThunk(
+    SmallVector<ValueDecl *, 1> stubClassLookupResults;
+    C.lookupInModule(M, ("$" + P->getNameStr()).str(), stubClassLookupResults);
+
+    if (stubClassLookupResults.size() > 0) {
```
I think we need to make sure that it's exactly one here via an ASSERT.

ktoso replied:
Weirdly enough I think i've seen multiple here... I'll see what I can do

xedin replied:
That won’t be unexpected though right? There should be only one $ thing per protocol?

ktoso replied:
Yeah not sure how multiple can happen, will dig into it

ktoso replied:
Seems I was seeing things, assert is fine -- added it back

xedin commented on lib/AST/ASTMangler.cpp:
```diff
@@ -4452,12 +4452,32 @@ void ASTMangler::appendDistributedThunk(
+    auto M = thunk->getModuleContext();
+
+    SmallVector<ValueDecl *, 1> stubClassLookupResults;
+    C.lookupInModule(M, ("$" + P->getNameStr()).str(), stubClassLookupResults);
```
I think you can use `llvm::Twine('$', P->getNameStr()).str()` might be slightly better here

ktoso replied:
Hm I think the + forms a Twine actually but I can be explicit about it yeap

xedin replied:
It would form std::string

ktoso replied:
Ah here I guess so, I misremembered some other piece of code I think. Thanks, fixed

xedin reviewed (approved):
LGTM! Could you add an example with array type and optional as parameters to make sure that it works as expected too?

ktoso commented:
Thanks for review! Huh seems I missed to git add my array test :) Adding some more

ktoso committed:
[Distributed] additional test coverage

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test
